### PR TITLE
Improve performance for pages with many toggl buttons + cleanups + todoist performance improvements

### PR DESCRIPTION
**Pull request by mgsloan**
## :star2: What does this PR do?

This PR makes some changes to `common.js`, along with minimal changes to `todoist.js` in order to mostly resolve the performance issues described in https://github.com/toggl/toggl-button/issues/1275

* Extracts out a function called `queryAndUpdateTimerLink` (better name suggestions appreciated).

* `queryAndUpdateTimerLink` is now run at the end of `renderTo` instead of at the end of `createTimerLink`. 
 
    - Here's why:  Code similar to `queryAndUpdateTimerLink` was being run after every single `createTimerLink` invocation.  This was tremendously inefficient, because it sends a message to the toggl-button background page asking it to say what the currently tracked task is.  The handler for every response to these calls would then scan every single link and modify it via activate / deactivate.  This is also discussed some [here](https://github.com/toggl/toggl-button/pull/1277#issuecomment-462013597).

    - This is still not really ideal, and probably runs `queryAndUpdateTimerLink` waaay more than should be necessary.

* Adds optional support for debouncing mutation observer results.  This is helpful if many mutations are expected to happen in rapid succession.  It also effectively defers rendering of toggl buttons until the page isn't as busy.

* Adds optional support for a specific target for the mutation observation, making it more efficient.

* Cleans up some code in `common.js`.

* Modifies the todoist integration to use mutation observer debouncing and targeting.

Note that it may be confusing to review this PR via "files changed", please take a look at individual commits.

## :bug: Recommendations for testing

Changes have been tested using the todoist integration in chrome.

## :memo: Links to relevant issues or information

* https://github.com/toggl/toggl-button/issues/1275

* https://github.com/toggl/toggl-button/pull/1277#issuecomment-462013597

### CODE REVIEW

File: src/scripts/common.js [hunk 1]
```diff
@@ -512,66 +502,58 @@ window.togglbutton = {
...
-
-    let link;
-
-    let i;
-
-    let minimal;
+    if (entry) {
+      for (const current of togglbutton.links) {
+        if (invokeIfFunction(current.params.description) !== entry.description) {
+          togglbutton.deactivateTimerLink(current.link);
```
Comment (shantanuraj) [inline src/scripts/common.js]: Is this required, since `activateTimerLink` at L524 will deactivate all other links anyway?

Comment (shantanuraj) [reply src/scripts/common.js]: If not, we could make it simpler by replacing L520-L529 with

```javascript
const current = togglbutton.links.find(button => invokeIfFunction(button.params.description) === entry.description);
togglbutton.activateTimerLink(current.link);
```

File: src/scripts/common.js [hunk 2]
```diff
@@ -118,20 +118,32 @@ window.togglbutton = {
...
-                return;
+              if (mutationSelector) {
+                const matches = mutations.filter(function (mutation) {
+                  return mutation.target.matches(mutationSelector);
+                });
+                if (!matches.length) {
+                  return;
+                }
+              }
+              if (opts.debounceObserve > 0) {
```
Comment (shantanuraj) [inline src/scripts/common.js]: `debounceInterval` might be a better choice for the option

File: src/scripts/common.js (hunk 1 above)
Comment (tcrammond) [reply src/scripts/common.js]: Adding in here that we've seen at least one integration in testing that now gets stuck in an infinite re-rendering loop (Atlassian Jira Cloud, issue page) when the button becomes "active". Wondering if it's related to the above *(I have not debugged it fully myself at this time, sorry)*.

File: src/scripts/common.js [hunk 3]
```diff
@@ -512,66 +502,58 @@ window.togglbutton = {
     // Add created link to links array
     togglbutton.links.push({ params: params, link: link });
 
-    // new button created - set state
+    return link;
+  },
+
+  // Query active timer entry, and set it to active. If a callback is
+  // provided as an argument, it will be invoked before updating the
```
Comment (tcrammond) [inline src/scripts/common.js]: It seems that a callback isn't an argument here in `queryAndUpdateTimerLink`, I assume it wasn't needed after all?

File: src/scripts/common.js (hunk 1 above)
Comment (mgsloan) [reply src/scripts/common.js]: :+1: Modified it to avoid the redundant deactivations and use `find`

**Comment by mgsloan** [comment]
@tcrammond Thanks, I appreciate your review!

**Pull request merged**